### ISSUE TITLE
hook problems with killing window

### ISSUE THREAD

**Issue by poetaman** [tmux/tmux#2849]
### Issue description

Let's say we create a session with a few windows. If we kill a window with kill-window command or by entering ``exit<CR>``. Both method have hook problems:

1. zsh autocompletion suggestion ``after-kill-window`` does not exist, it gives a invalid option error.
2. When killing with ``exit<CR>``, pane-focus-in is triggered as focus jumps to previous window, but no after-kill-pane or after-kill-window.
3. When killing with ``kill-window``, pane-focus-out is triggered with invalid ``#{session_name}`` for correct hook_pane (session still exists so not sure why empty session_name). pane-focus-in is triggered too, and seems correct. No ``after-kill-pane`` or ``after-kill-window`` triggered.

### Required information

Please provide the following information:

* tmux version (`tmux -V`) next-3.3.
* Platform (`uname -sp`) Darwin arm.
* $TERM inside and outside of tmux (`echo $TERM`) xterm-256color, tmux-256color.

**Comment by nicm** [comment]
Instead of opening an issue for every hook, why don't you look at the man page and make one list of hooks you think are missing or are being fired in the wrong way?

**Comment by reportaman** [comment]
@nicm Some opensource authors don't like 1 ticket with multiple tasks/problems in it. Got it...

**Comment by nicm** [comment]
Multiple different problems, sure, but you are just asking about basically the same thing. I don't mind changing or adding hooks but there is no point in having them all individually if you have many suggestions.

**Comment by reportaman** [comment]
@nicm Read point 3. of this ticket... its unique to killing window, so not the same as my other ticket which focussed on killing pane. Why would session name be empty for pane-focus-out triggered while killing window?

**Comment by nicm** [comment]
Did killing the window also kill the session?

**Comment by reportaman** [comment]
@nicm No, there are multiple other windows still remaining in the session. The focus just moved to another window of the same session.

**Comment by nicm** [comment]
I'll have to take a look.

**Comment by nicm** [comment]
Are all these points still valid? What changes would you like to see here?

**Comment by reportaman** [comment]
@nicm Point 1,3 are still valid, 2 was my mistake in understanding while filing this ticket.

1) Try setting a hook for ``after-kill-window``, tmux will give error. Such hook doesn't exist, which from the documentation seems like it would exist. ``after-kill-pane`` exists and is triggered.

2) Ignore this, the pane exited with ``exit<CR>`` triggers pane-exited, and has nothing to do with after-kill-pane or after-kill-window

3) For pane-focus-out triggered due to kill-window, the hook_pane is the pane_id of pane that just got killed, and for that pane_id, the session_name is empty when printed from within the hook using ``-t <killed_pane_id>``.

When we provide a hook_pane, the idea is that using that user can find other information about that pane by using it as the value passed to target option ``-t``. That seems to be not true if the pane just got killed.

**Comment by nicm** [comment]
For (1), yeah, `after-kill-window` does not exist. I think we can probably add it, although `window-unlinked` seems more generally useful. I think it was probably left out because you don't have access to the window anymore when it runs, but other hooks have the same problem: [killw.diff.txt](https://github.com/user-attachments/files/23256797/killw.diff.txt)

For (3), obviously you can't use use the killed pane ID after it has been killed, so the session name will be empty. The easiest thing to do might be to set `hook_window` as well and then you could use it: [hook_win.diff.txt](https://github.com/user-attachments/files/23256880/hook_win.diff.txt)
